// @dicy/core/lib/types.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
const DEFAULT_OPTIONS = exports.DEFAULT_OPTIONS = { $BIBINPUTS: ['$ROOTDIR', '$ROOTDIR/$OUTDIR', ''],
  $TEXINPUTS: ['$ROOTDIR', '$ROOTDIR/$OUTDIR', ''],
  bibtexEngine: 'bibtex',
  cleanPatterns: ['$OUTDIR/$JOB!($OUTEXT|.synctex.gz|.tex)', '/$OUTDIR/_minted-$JOB/*'],
  disableRules: [],
  dviToPdfEngine: 'xdvipdfmx',
  engine: 'pdflatex',
  epstopdfBoundingBox: 'default',
  epstopdfOutputPath: '$DIR_0/$NAME_0.pdf',
  indexAutomaticRanges: true,
  indexEngine: 'makeindex',
  indexOrdering: 'word',
  indexSorting: 'default',
  knitrConcordance: true,
  knitrOutputPath: '$JOB.tex',
  lhs2texStyle: 'poly',
  literateAgdaEngine: 'agda',
  literateHaskellEngine: 'lhs2TeX',
  outputFormat: 'pdf',
  phaseCycles: 20,
  pweaveCacheDirectory: 'pweave-cache-for-$JOB',
  pweaveFigureDirectory: 'pweave-figures-for-$JOB',
  pweaveOutputFormat: 'tex',
  pweaveOutputPath: '$JOB.tex',
  severity: 'warning'

  // END_AUTO

};

// START_AUTO
