// @dicy/core/lib/Rules/SaveCache.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _set = require('babel-runtime/core-js/set');

var _set2 = _interopRequireDefault(_set);

var _from = require('babel-runtime/core-js/array/from');

var _from2 = _interopRequireDefault(_from);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _File = require('../File');

var _File2 = _interopRequireDefault(_File);

var _Rule = require('../Rule');

var _Rule2 = _interopRequireDefault(_Rule);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class SaveCache extends _Rule2.default {

  initialize() {
    var _this = this;

    return (0, _asyncToGenerator3.default)(function* () {
      _this.cacheFilePath = _this.resolvePath('$ROOTDIR/$NAME-cache.yaml');
    })();
  }

  preEvaluate() {
    var _this2 = this;

    return (0, _asyncToGenerator3.default)(function* () {
      // If all output files are virtual the don't bother saving.
      if ((0, _from2.default)(_this2.rules).every(function (rule) {
        return rule.outputs.every(function (file) {
          return file.virtual;
        });
      })) {
        _this2.actions.delete('run');
      }
    })();
  }

  run() {
    var _this3 = this;

    return (0, _asyncToGenerator3.default)(function* () {
      const cache = {
        filePath: _this3.filePath,
        options: _this3.state.options,
        files: {},
        rules: []

        // Loop through all the files and add them to the cache.
      };for (const file of _this3.files) {
        const fileCache = {
          timeStamp: file.timeStamp,
          hash: file.hash,
          value: file.value,
          jobNames: (0, _from2.default)(file.jobNames.values())
        };

        if (file.type) {
          fileCache.type = file.type;
        }

        if (file.subType) {
          fileCache.subType = file.subType;
        }

        cache.files[file.filePath] = fileCache;
      }

      // Loop through all the rules and add them to the cache.
      for (const rule of _this3.rules) {
        const ruleCache = {
          name: rule.constructor.name,
          command: rule.command,
          phase: rule.phase,
          parameters: rule.parameters.map(function (file) {
            return file.filePath;
          }),
          inputs: rule.inputs.map(function (file) {
            return file.filePath;
          }),
          outputs: rule.outputs.map(function (file) {
            return file.filePath;
          })
        };

        if (rule.options.jobName) {
          ruleCache.jobName = rule.options.jobName;
        }

        cache.rules.push(ruleCache);
      }

      // Save the cache and update the timestamp.
      yield _File2.default.safeDump(_this3.cacheFilePath, cache);
      _this3.state.cacheTimeStamp = yield _File2.default.getModifiedTime(_this3.cacheFilePath);

      return true;
    })();
  }
}
exports.default = SaveCache;
SaveCache.commands = new _set2.default(['save']);
SaveCache.alwaysEvaluate = true;
SaveCache.ignoreJobName = true;
SaveCache.description = 'Saves file and rule status to a cache (-cache.yaml) to assist with rebuilding.';
